lib/tesla/middleware/telemetry.ex

```diff
@@ -13,21 +13,102 @@ if Code.ensure_loaded?(:telemetry) do
 
     end
 
-    :telemetry.attach("my-tesla-telemetry", [:tesla, :request], fn event, time, meta, config ->
+    :telemetry.attach("my-tesla-telemetry", [:tesla, :request, :stop], fn event, measurements, meta, config ->
       # Do something with the event
     end)
     ```
 
+    ## Options
+
+    * `:event_prefix` - a list of atoms to prefix to the telemetry event name. This can be set if you need to distinguish events from different clients. Defaults to `[]`
+
+    ## Telemetry Events
+
+    * `[:tesla, :request, :start]` - emitted at the beginning of the request.
+      * Measurement: `%{time: System.monotonic_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :stop]` - emitted at the end of the request.
+      * Measurement: `%{duration: native_time}`
+      * Metadata: `%{env: Tesla.Env.t}`
+
+    * `[:tesla, :request, :error]` - emitted when there is an error.
+      * Measurement: `%{value: 1}`
+      * Metadata: `%{env: Tesla.Env.t, kind: Exception.kind | nil, reason: term, stacktrace: Exception.stacktrace}`
+
+    ## Legacy Telemetry Events
+
+      * `[:tesla, :request]` - This event is emitted for backwards compatibility only and should be considered deprecated.
+      This event can be disabled by setting `config :tesla, Tesla.Middleware.Telemetry, disable_legacy_event: true` in your config. Be sure to run `mix deps.compile --force tesla` after changing this setting to ensure the change is picked up.
+
     Please check the [telemetry](https://hexdocs.pm/telemetry/) for the further usage.
     """
 
+    @disable_legacy_event Application.get_env(:tesla, Tesla.Middleware.Telemetry,
+                            disable_legacy_event: false
+                          )[:disable_legacy_event]
+
     @behaviour Tesla.Middleware
 
     @impl Tesla.Middleware
-    def call(env, next, _opts) do
-      {time, res} = :timer.tc(Tesla, :run, [env, next])
-      :telemetry.execute([:tesla, :request], %{request_time: time}, %{result: res})
-      res
+    def call(env, next, opts) do
+      prefix = Keyword.get(opts, :event_prefix, [])
+      start_time = System.monotonic_time()
+
+      emit_start(env, start_time, prefix)
+
+      try do
+        Tesla.run(env, next)
+      catch
+        kind, reason ->
+          stacktrace = System.stacktrace()
+          metadata = %{env: env, kind: kind, reason: reason, stacktrace: stacktrace}
+
+          :telemetry.execute(
+            prefix ++ [:tesla, :request, :error],
+            %{value: 1},
+            metadata
+          )
+
+          emit_stop(env, start_time, prefix, {:error, reason})
```

Comment:
In this case we emit `:error` and no `:stop` but in the error tuple below, we emit both `:error` and `:stop`. I think we the consensus was to always emit `:stop`.

Reply:
The error is emitted on line 67. The `emit_stop/4` with that last argument is only there for backward compatibility with the previous implementation, which emitted the result tuple as metadata. I was creating an error tuple to pass to it to fit the previous pattern.

I wonder if it may be more useful to just be more explicitly verbose and mark the deprecated pieces to be deleted later in comments rather than this function.